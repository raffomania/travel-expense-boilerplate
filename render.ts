import { parse as parseYaml } from "https://deno.land/std@0.63.0/encoding/yaml.ts";
import { prop, sum } from "https://deno.land/x/ramda@v0.27.2/mod.ts";

export var SMALL_FOOD_MONEY = 14;
export var BIG_FOOD_MONEY = 28;

export interface RawInput {
    city: string;
    date?: string;
    title: string;
    author: string;
    start_date: string;
    end_date: string;
    company: string;
    signature_image?: string;
    entries: ProcessedEntry[];
}

export interface ProcessedInput extends RawInput {
    date: string;
    total_km: number;
    total_driving_costs: number;
    total_food_money: number;
}

export interface RawEntry {
    date: string;
    start_time: string;
    subject: string;
    hours: number;
    km: number;
    end_time: string;
    food_money?: number;
}

export interface ProcessedEntry {
    date: string;
    start_time: string;
    subject: string;
    hours: number;
    km: number;
    food_money: number;
    end_time: string;
}

export async function load(): Promise<RawInput> {
    const decoder = new TextDecoder("utf-8");

    const yamlFile = await Deno.readFile(Deno.args[0]);

    /** Decoding the file text. */
    const yamlText = decoder.decode(yamlFile);

    /** Returning parsed yaml as an object. */
    return (await parseYaml(yamlText)) as RawInput;
}

export function process(rawDetails: RawInput): ProcessedInput {
    const entries = rawDetails.entries.map(processEntry);

    const total_km = sum(entries.map(prop("km")));
    const total_driving_costs = total_km * 0.3;
    const total_food_money = sum(entries.map(prop("food_money")));

    const today = new Date();
    const date =
        rawDetails.date ??
        `${today.getDate()}.${today.getMonth() + 1}.${today.getFullYear()}`;

    return {
        ...rawDetails,
        total_km,
        total_driving_costs,
        total_food_money,
        entries,
        date,
    };
}

function processEntry(raw: RawEntry): ProcessedEntry {
    // Set some default in case those values aren't specified
    if (raw.km == undefined) {
        raw.km = 0;
    }
    if (raw.start_time == undefined) {
        raw.start_time = "00:00";
    }
    if (raw.end_time == undefined) {
        raw.end_time = "24:00";
    }

    // Autocalculate hours unless explicitly defined.
    if (raw.hours == undefined) {
        // Calculate the hours based on the given time strings.
        // Let the javascript dateparser do its job.
        var start = new Date(`01/01/2024 ${raw.start_time}`);
        var end = new Date(`01/01/2024 ${raw.end_time}`);

        var milliseconds_diff = end - start;
        raw.hours = Math.floor(milliseconds_diff / 1000 / 60 / 60);
    }

    // Autocalculate food_money unless explicitly defined.
    if (raw.food_money == undefined) {
        if (raw.hours >= 8 && raw.hours < 24) {
            raw.food_money = SMALL_FOOD_MONEY;
        } else if (raw.hours >= 24) {
            raw.food_money = BIG_FOOD_MONEY;
        } else {
            raw.food_money = 0;
        }
    }

    return {
        food_money: raw.food_money!,
        ...raw,
    };
}
