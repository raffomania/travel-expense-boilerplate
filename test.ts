import { assertSnapshot } from "https://deno.land/std@0.224.0/testing/snapshot.ts";
import { parse as parseYaml } from "https://deno.land/std@0.63.0/encoding/yaml.ts";
import render from "./template.ts";
import { RawInput, process } from "./render.ts";

// These tests ensure that old yaml data renders the same as before,
// even when their data differs from values that are now automatically
// calculated.
Deno.test("basicSnapshot", async function (t): Promise<void> {
    const yaml = process(
        (await parseYaml(`
author: Herbert Derpert
company: Corp
date: 10.05.2024
start_date: 01.04.2024
end_date: 30.04.2023
city: Hamburg
title: Travelin

entries:
    - index: 1
      date: 10.04.
      start_time: "08:30"
      end_time: "17:30"
      subject: "Did some stuff"
      hours: 9
      km: 20
      food_money: 14
    - index: 2
      date: 23.04.
      start_time: "10:00"
      end_time: "18:00"
      subject: "There's more stuff here"
      hours: 8
      km: 20
      food_money: 10
    `)) as RawInput
    );
    const rendered = render(yaml);
    await assertSnapshot(t, rendered);
});

Deno.test("snapshotNoIndices", async function (t): Promise<void> {
    const yaml = process(
        (await parseYaml(`
author: Herbert Derpert
company: Corp
date: 10.05.2024
start_date: 01.04.2024
end_date: 30.04.2023
city: Hamburg
title: Travelin

entries:
    - date: 10.04.
      start_time: "08:30"
      end_time: "17:30"
      subject: "Did some stuff"
      hours: 9
      km: 20
      food_money: 14
    - date: 23.04.
      start_time: "10:00"
      end_time: "18:00"
      subject: "There's more stuff here"
      hours: 8
      km: 20
      food_money: 10
    `)) as RawInput
    );
    const rendered = render(yaml);
    await assertSnapshot(t, rendered);
});

Deno.test("snapshotNoFoodMoney", async function (t): Promise<void> {
    const yaml = process(
        (await parseYaml(`
author: Herbert Derpert
company: Corp
date: 10.05.2024
start_date: 01.04.2024
end_date: 30.04.2023
city: Hamburg
title: Travelin

entries:
    - date: 10.04.
      start_time: "08:30"
      end_time: "17:30"
      subject: "Did some stuff"
      hours: 9
      km: 20
    - date: 23.04.
      start_time: "10:00"
      end_time: "18:00"
      subject: "There's more stuff here"
      hours: 20
      km: 20
    `)) as RawInput
    );
    const rendered = render(yaml);
    await assertSnapshot(t, rendered);
});

Deno.test("snapshotNoHours", async function (t): Promise<void> {
    const yaml = process(
        (await parseYaml(`
author: Herbert Derpert
company: Corp
date: 10.05.2024
start_date: 01.04.2024
end_date: 30.04.2023
city: Hamburg
title: Travelin

entries:
    - date: 10.04.
      start_time: "08:30"
      end_time: "17:30"
      subject: "Did some stuff"
      km: 20
    - date: 23.04.
      start_time: "10:00"
      end_time: "18:00"
      subject: "There's more stuff here"
      km: 20
    `)) as RawInput
    );
    const rendered = render(yaml);
    await assertSnapshot(t, rendered);
});
