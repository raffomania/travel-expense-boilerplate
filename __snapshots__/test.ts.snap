export const snapshot = {};

snapshot[`basicSnapshot 1`] = `
\`
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title></title>

  <style>* {
    box-sizing: border-box;
}
body {
    font-family: Noto Sans;
    font-size: 12pt;
    padding: 2em;
}
th {
    font-size: 10pt;
    text-align: left;
}
table,
tr,
td {
    border: 1px solid #888;
    border-collapse: collapse;
}
th,
td {
    padding: 0.5em;
}

img.signature {
    max-width: 10em;
}
</style>
</head>
<body>
  <header id="title-block-header">
  <span style="float: right;">Vom 01.04.2024 bis 30.04.2023</span>
  <h1 class="title">Travelin</h1>
  <p class="author">Name: Herbert Derpert, Firma: Corp</p>
  </header>

  <table style="width: 100%;" class="pure-table pure-table-bordered">
    <thead>
      <tr>
        <th>lfd. Nr.</th>
        <th>Datum</th>
        <th>Reisebeginn/-ende</th>
        <th>Reiseanlass; Reiseweg (Ziel und Zweck der Reise)</th>
        <th>Std.</th>
        <th>dienstl. gefahrene km</th>
        <th>Verpflegungspauschbeträge, mind. 8h = 14,00 €, 24h = 28,00 €</th>
      </tr>
    </thead>

    <tbody>
      
<tr valign="top">
  <td rowspan=2>1</td>
  <td rowspan=2>10.04.</td>
  <td>08:30</td>
  <td rowspan=2>Did some stuff</td>
  <td rowspan=2>9</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>14,00 €</td>
</tr>
<tr>
  <td>17:30</td>
</tr>
  
<tr valign="top">
  <td rowspan=2>2</td>
  <td rowspan=2>23.04.</td>
  <td>10:00</td>
  <td rowspan=2>There's more stuff here</td>
  <td rowspan=2>8</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>10,00 €</td>
</tr>
<tr>
  <td>18:00</td>
</tr>
  
      <tr valign="top">
        <td colspan=5 align="right"><strong>Summen</strong></td>
        <td>40 km x 0,30 € = <strong>12,00 €</strong></td>
        <td><strong>24,00 €</strong></td>
      </tr>
    </tbody>
  </table>

  <p><strong>Gesamtsumme der Reisekosten: 36,00 €</strong></p>

  <p>Hamburg, den 10.05.2024</p>
  <p>Herbert Derpert</p>
  <img class="signature" src="data:image/png;base64,"/>
</body>
</html>\`
`;

snapshot[`snapshotNoIndices 1`] = `
\`
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title></title>

  <style>* {
    box-sizing: border-box;
}
body {
    font-family: Noto Sans;
    font-size: 12pt;
    padding: 2em;
}
th {
    font-size: 10pt;
    text-align: left;
}
table,
tr,
td {
    border: 1px solid #888;
    border-collapse: collapse;
}
th,
td {
    padding: 0.5em;
}

img.signature {
    max-width: 10em;
}
</style>
</head>
<body>
  <header id="title-block-header">
  <span style="float: right;">Vom 01.04.2024 bis 30.04.2023</span>
  <h1 class="title">Travelin</h1>
  <p class="author">Name: Herbert Derpert, Firma: Corp</p>
  </header>

  <table style="width: 100%;" class="pure-table pure-table-bordered">
    <thead>
      <tr>
        <th>lfd. Nr.</th>
        <th>Datum</th>
        <th>Reisebeginn/-ende</th>
        <th>Reiseanlass; Reiseweg (Ziel und Zweck der Reise)</th>
        <th>Std.</th>
        <th>dienstl. gefahrene km</th>
        <th>Verpflegungspauschbeträge, mind. 8h = 14,00 €, 24h = 28,00 €</th>
      </tr>
    </thead>

    <tbody>
      
<tr valign="top">
  <td rowspan=2>1</td>
  <td rowspan=2>10.04.</td>
  <td>08:30</td>
  <td rowspan=2>Did some stuff</td>
  <td rowspan=2>9</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>14,00 €</td>
</tr>
<tr>
  <td>17:30</td>
</tr>
  
<tr valign="top">
  <td rowspan=2>2</td>
  <td rowspan=2>23.04.</td>
  <td>10:00</td>
  <td rowspan=2>There's more stuff here</td>
  <td rowspan=2>8</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>10,00 €</td>
</tr>
<tr>
  <td>18:00</td>
</tr>
  
      <tr valign="top">
        <td colspan=5 align="right"><strong>Summen</strong></td>
        <td>40 km x 0,30 € = <strong>12,00 €</strong></td>
        <td><strong>24,00 €</strong></td>
      </tr>
    </tbody>
  </table>

  <p><strong>Gesamtsumme der Reisekosten: 36,00 €</strong></p>

  <p>Hamburg, den 10.05.2024</p>
  <p>Herbert Derpert</p>
  <img class="signature" src="data:image/png;base64,"/>
</body>
</html>\`
`;

snapshot[`snapshotNoFoodMoney 1`] = `
\`
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title></title>

  <style>* {
    box-sizing: border-box;
}
body {
    font-family: Noto Sans;
    font-size: 12pt;
    padding: 2em;
}
th {
    font-size: 10pt;
    text-align: left;
}
table,
tr,
td {
    border: 1px solid #888;
    border-collapse: collapse;
}
th,
td {
    padding: 0.5em;
}

img.signature {
    max-width: 10em;
}
</style>
</head>
<body>
  <header id="title-block-header">
  <span style="float: right;">Vom 01.04.2024 bis 30.04.2023</span>
  <h1 class="title">Travelin</h1>
  <p class="author">Name: Herbert Derpert, Firma: Corp</p>
  </header>

  <table style="width: 100%;" class="pure-table pure-table-bordered">
    <thead>
      <tr>
        <th>lfd. Nr.</th>
        <th>Datum</th>
        <th>Reisebeginn/-ende</th>
        <th>Reiseanlass; Reiseweg (Ziel und Zweck der Reise)</th>
        <th>Std.</th>
        <th>dienstl. gefahrene km</th>
        <th>Verpflegungspauschbeträge, mind. 8h = 14,00 €, 24h = 28,00 €</th>
      </tr>
    </thead>

    <tbody>
      
<tr valign="top">
  <td rowspan=2>1</td>
  <td rowspan=2>10.04.</td>
  <td>08:30</td>
  <td rowspan=2>Did some stuff</td>
  <td rowspan=2>9</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>14,00 €</td>
</tr>
<tr>
  <td>17:30</td>
</tr>
  
<tr valign="top">
  <td rowspan=2>2</td>
  <td rowspan=2>23.04.</td>
  <td>10:00</td>
  <td rowspan=2>There's more stuff here</td>
  <td rowspan=2>20</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>14,00 €</td>
</tr>
<tr>
  <td>18:00</td>
</tr>
  
      <tr valign="top">
        <td colspan=5 align="right"><strong>Summen</strong></td>
        <td>40 km x 0,30 € = <strong>12,00 €</strong></td>
        <td><strong>28,00 €</strong></td>
      </tr>
    </tbody>
  </table>

  <p><strong>Gesamtsumme der Reisekosten: 40,00 €</strong></p>

  <p>Hamburg, den 10.05.2024</p>
  <p>Herbert Derpert</p>
  <img class="signature" src="data:image/png;base64,"/>
</body>
</html>\`
`;

snapshot[`snapshotNoHours 1`] = `
\`
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title></title>

  <style>* {
    box-sizing: border-box;
}
body {
    font-family: Noto Sans;
    font-size: 12pt;
    padding: 2em;
}
th {
    font-size: 10pt;
    text-align: left;
}
table,
tr,
td {
    border: 1px solid #888;
    border-collapse: collapse;
}
th,
td {
    padding: 0.5em;
}

img.signature {
    max-width: 10em;
}
</style>
</head>
<body>
  <header id="title-block-header">
  <span style="float: right;">Vom 01.04.2024 bis 30.04.2023</span>
  <h1 class="title">Travelin</h1>
  <p class="author">Name: Herbert Derpert, Firma: Corp</p>
  </header>

  <table style="width: 100%;" class="pure-table pure-table-bordered">
    <thead>
      <tr>
        <th>lfd. Nr.</th>
        <th>Datum</th>
        <th>Reisebeginn/-ende</th>
        <th>Reiseanlass; Reiseweg (Ziel und Zweck der Reise)</th>
        <th>Std.</th>
        <th>dienstl. gefahrene km</th>
        <th>Verpflegungspauschbeträge, mind. 8h = 14,00 €, 24h = 28,00 €</th>
      </tr>
    </thead>

    <tbody>
      
<tr valign="top">
  <td rowspan=2>1</td>
  <td rowspan=2>10.04.</td>
  <td>08:30</td>
  <td rowspan=2>Did some stuff</td>
  <td rowspan=2>9</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>14,00 €</td>
</tr>
<tr>
  <td>17:30</td>
</tr>
  
<tr valign="top">
  <td rowspan=2>2</td>
  <td rowspan=2>23.04.</td>
  <td>10:00</td>
  <td rowspan=2>There's more stuff here</td>
  <td rowspan=2>8</td>
  <td rowspan=2>20 km x 0,30 €</td>
  <td rowspan=2>14,00 €</td>
</tr>
<tr>
  <td>18:00</td>
</tr>
  
      <tr valign="top">
        <td colspan=5 align="right"><strong>Summen</strong></td>
        <td>40 km x 0,30 € = <strong>12,00 €</strong></td>
        <td><strong>28,00 €</strong></td>
      </tr>
    </tbody>
  </table>

  <p><strong>Gesamtsumme der Reisekosten: 40,00 €</strong></p>

  <p>Hamburg, den 10.05.2024</p>
  <p>Herbert Derpert</p>
  <img class="signature" src="data:image/png;base64,"/>
</body>
</html>\`
`;
